### PR TITLE
feat(mysql): close idling connections in pool

### PR DESCRIPTION
**Pull request by chenba**
Because:
 - the mysql module's connection pool never close any of its connections

This commit:
 - close a connection after it idles for a configurable amount of time

### CODE REVIEW

File: packages/fxa-shared/db/mysql.ts [hunk 1]
```diff
@@ -251,6 +259,18 @@ export class MysqlStoreShared {
...
 
+        // The mysql module's connection pool never closes any of its
+        // connections.  There's at least one open contributor PR on the
+        // module's repo for the feature, but not much activity.
+        if (conn._fxa_idle_timeout) {
+          clearTimeout(conn._fxa_idle_timeout);
+        }
+
+        conn._fxa_idle_timeout = setTimeout(() => {
+          (pool as any)._purgeConnection(conn);
```
Comment (dschom) [inline packages/fxa-shared/db/mysql.ts:270]: There are cases where the call to _purgeConnection can throw an error. Since this is pretty deep in the code, and happens in a setTimeout, should there be some error handling?

**Review by dschom — COMMENTED** [review]
It looks pretty good, but this also seems like a delicate and legacy piece of code, so I am being careful with approval.

File: packages/fxa-shared/test/db/mysql.ts [hunk 2]
```diff
@@ -143,4 +158,18 @@ describe('mysql', function () {
       dbOpts.connectionLimit * 2
     );
   });
+
+  it('closes connections after idling', async () => {
```
[the remark this answers is not in the document]
Comment (chenba) [reply packages/fxa-shared/test/db/mysql.ts:162]: The code is in the callback for getting a connection from the pool, so we shouldn't be getting a connection that's already been released?

> Also can we hold reference to the connection and inspect its state?

Do you mean in the tests?

File: packages/fxa-shared/db/mysql.ts (hunk 1 above)
Comment (chenba) [reply packages/fxa-shared/db/mysql.ts:270]: A fine idea.  How do you feel about sending a StatsD stat and writing the error to the log?  There isn't any recovery we can do in here.

[the remark this answers is not in the document]
Comment (chenba) [reply packages/fxa-shared/db/mysql.ts:270]: [The `connection.realEnd` call](https://github.com/mysqljs/mysql/blob/dc9c152a87ec51a1f647447268917243d2eab1fd/lib/Pool.js#L261) in `_purgeConnection` will lead to `connection.destroy` being called (though admittedly in a roundabout way).  _purgeConnection removes the the connection from the pool first, which is something we'd want.

Comment (dschom) [reply packages/fxa-shared/db/mysql.ts:270]: A statsd call would be great.

[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-shared/db/mysql.ts:270]: Okay sounds good. I see what you are saying about realEnd now. 

Mostly I just wanted to make sure the connection is destroyed. I think validating this in the test, would be good. There is just enough state here that I can see things going sideways, or a change to an internal routine in this package in the future causing a problem.

File: packages/fxa-shared/test/db/mysql.ts [hunk 3]
```diff
@@ -162,14 +163,19 @@ describe('mysql', function () {
...
     await new Promise((resolve) => {
       setTimeout(resolve, 50);
     });
     const closeConnCount = statsdStub.increment.args.filter(
-      (x) => x[0] === 'mysql.close_connection'
+      (x) => x[0] === 'mysql.close_idle_connection'
     ).length;
     assert.equal(closeConnCount, 2);
+    assert.equal((conn1 as Connection).state, 'disconnected');
+    assert.equal((conn2 as Connection).state, 'disconnected');
```
Comment (chenba) [inline packages/fxa-shared/test/db/mysql.ts:179]: @dschom is this what you meant?

File: packages/fxa-shared/test/db/mysql.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dschom) [reply packages/fxa-shared/test/db/mysql.ts:163]: > so we shouldn't be getting a connection that's already been released?

Sorry my comment was a bit vague. You are correct the incoming connection will always be in a released state, but what is the state of the connection when the idle timeout is hit and the connection is purged? 

If the connection has already been released, and it could be since each connection [is released](https://github.com/mozilla/fxa/blob/69baebee8b3d38c444bff0fa5274f16412c0fd9f/packages/fxa-shared/db/mysql.ts#L226) after the query ends. Then when the timeout triggers you will purge an already released connection, which seems like it would result in an error as the following code paths are executed:

```
pool._purgeConnection(conn)
|-> this._removeConnection(conn)
|-> this.releaseConnection(conn)
      |-> throw new Error('Connection already released'); //Pool.js#L138
```

I have a feeling you are more familiar with the code than me, so perhaps this edge cases is bogus?

> Do you mean in the tests?

Yes in the test. Just to make sure the purge operation, which is internal, really had and continues to have the desired effect.

File: packages/fxa-shared/test/db/mysql.ts (hunk 3 above)
Comment (dschom) [reply packages/fxa-shared/test/db/mysql.ts:179]: Yep! 👍

File: packages/fxa-shared/test/db/mysql.ts [hunk 4]
```diff
@@ -143,4 +159,23 @@ describe('mysql', function () {
...
     );
   });
+
+  it('closes connections after idling', async () => {
+    // pass a lower idle limit config value
+    mysql = new MySqlStoreTest({ ...dbOpts, idleLimitMs: 10 });
+    const conn1 = await mysql.getConnection();
+    const conn2 = await mysql.getConnection();
+    assert.equal((conn1 as Connection).state, 'connected');
+    assert.equal((conn2 as Connection).state, 'connected');
```
Comment (dschom) [inline packages/fxa-shared/test/db/mysql.ts:169]: It seems like this can vary between authenticated and connected. Maybe it is not neccessary to check here?

Comment (chenba) [reply packages/fxa-shared/test/db/mysql.ts:169]: Good point.  I can assert that it is not 'disconnected', which is the initial state.

**Review by dschom — APPROVED** [review]
LGTM! 📉 

Thanks for entertaining all my silly questions.
